test/test_nn.py

```diff
@@ -3410,63 +3405,89 @@ def _multihead_attn_test_helper(add_key_padding_mask, add_bias_kv=False, add_zer
+                        multihead_attn_module.add_zero_attn, multihead_attn_module.dropout,
+                        multihead_attn_module.out_proj.weight, multihead_attn_module.out_proj.bias,
+                        multihead_attn_module.training, key_padding_mask_tensor, True, attn_mask_tensor,
+                        static_k=saved_k_tensor, static_v=saved_v_tensor) 
+                else:
+                    result, result_weight = torch.nn.functional.multi_head_attention_forward(
+                        _Q, _K, _V,
+                        d_model, nheads,
+                        multihead_attn_module.in_proj_weight, multihead_attn_module.in_proj_bias,
+                        multihead_attn_module.bias_k, multihead_attn_module.bias_v,
+                        multihead_attn_module.add_zero_attn, multihead_attn_module.dropout,
+                        multihead_attn_module.out_proj.weight, multihead_attn_module.out_proj.bias,
+                        multihead_attn_module.training, key_padding_mask_tensor, True, attn_mask_tensor,
+                        True, multihead_attn_module.q_proj_weight, 
+                        multihead_attn_module.k_proj_weight, multihead_attn_module.v_proj_weight,
+                        static_k=saved_k_tensor, static_v=saved_v_tensor) 
 
                 result = result.squeeze(0).detach().numpy()
 
-                Q_fc = _fc(Q, "in_proj_", multihead_attn_module, end=d_model)
-                K_fc = _fc(
-                    K, "in_proj_", multihead_attn_module, start=d_model, end=2 * d_model
-                )
-                V_fc = _fc(V, "in_proj_", multihead_attn_module, start=2 * d_model)
+                q_proj_weight = multihead_attn_module.in_proj_weight[:d_model]
+                k_proj_weight = multihead_attn_module.in_proj_weight[d_model:(d_model * 2)]
+                v_proj_weight = multihead_attn_module.in_proj_weight[(d_model * 2):]
+                if not multihead_attn_module._qkv_same_embed_dim:
+                    q_proj_weight = multihead_attn_module.q_proj_weight
+                    k_proj_weight = multihead_attn_module.k_proj_weight
+                    v_proj_weight = multihead_attn_module.v_proj_weight
+
+                Q_fc = _fc(Q, q_proj_weight, multihead_attn_module.in_proj_bias[:d_model])
+                K_fc = _fc(K, k_proj_weight, multihead_attn_module.in_proj_bias[d_model:(d_model * 2)])
+                V_fc = _fc(V, v_proj_weight, multihead_attn_module.in_proj_bias[(d_model * 2):])
 
                 if add_bias_kv:
                     K_fc = np.concatenate((K_fc, np.repeat(bias_k, K_fc.shape[0], axis=0)), axis=1)
                     V_fc = np.concatenate((V_fc, np.repeat(bias_v, V_fc.shape[0], axis=0)), axis=1)
-                    attn_mask = np.concatenate((attn_mask, np.ones([1, 1])), axis=1)
+                    if attn_mask is not None:
+                        attn_mask = np.concatenate((attn_mask, np.ones([1, 1])), axis=1)
```

Comment:
What are these `np` functions everywhere and how will they work with cuda tensors?

Reply:
they are used to re-write the multiheadattention function and verify the results of nn.MulitheadAttention module.